Parsing xml files with slightly differing spec of ImageLoader format "bdv.multimg.zarr"

m-albert opened the pull request:
Hi @d-v-b , I finally played around a bit with this library :)

I was trying to parse the xml file of the dataset shared by @nvladimus here: https://forum.image.sc/t/are-there-publicly-available-mesospim-datasets-unprocessed/116020/9.

There were some things I got a bit stuck with, as the xml seems to be following a slightly different spec from the test files in this repo (it does seem to load just fine in BigStitcher / BigDataViewer). Notably the following:

### 1) zgroup

A zgroup occurs as

```xml
<zgroup setup="0" tp="0" path="s0-t0.zarr" indicies="0 0" />
```
instead of e.g.
```xml
<zgroup setup="0" timepoint="0">
  <path>tile_x_0000_y_0000_z_0000_ch_488.zarr</path>
</zgroup>
```
Not modeled differences that cause validation errors:
1. "tp" instead of "timepoint"
1. "path" is an attribute instead of an element.

Both "tp" and "path" are read with their specific names by BigDataViewer. Which I'm actually not understanding as https://github.com/bigdataviewer/bigdataviewer-omezarr seems to expect differently. But I'm not very familiar with Java and might be looking at the wrong place and missing things...

### 2) Additional elements

E.g.:
- `/SpimData/generatedBy`
- `/SpimData/SequenceDescription/ViewSetups/ViewSetup/camera`

These create validation errors as well (because of changing the xml order?).

### Question

What would you say would be a good way to deal with these differences while still making use of the nice pydantic models in this library?

I made some modifications [here](https://github.com/d-v-b/pydantic-bigstitcher/compare/main...m-albert:pydantic-bigstitcher:mesospim) until parsing worked fine, but have very limited experience with bigstitcher xml files (as well as parsing and validation using pydantic).

Thanks a lot for any comments / help 🙏

EDIT: I originally wanted to open an issue but accidentally created a PR 🙈. Happy to move it if you prefer!

d-v-b commented:
hi @m-albert, my observation has been that bigstitcher xml documents change their schema without warning as new versions are released. So it's very likely that Nikita's data was processed with a different version of bigstitcher than what I wrote this code to support (the allen institute's data pipeline).

Generally the way to handle this on the pydantic side is to define certain fields as unions over the various possibilities:

```python

class ZGroupA(BaseXmlModel, tag="zgroup"):
    setup: str = attr()
    timepoint: str = attr()
    path: str | None = element(default=None)

class ZGroupB(BaseXmlModel, tag="zgroup"):
    setup: str = attr()
    tp: str = attr()
    path: str | None = element(default=None)
   indices: WhiteSpaceSeparatedStringBecauseArraysDon'tExistLOL
```

Operationally, I would recommend adding the full xml document to the set of [test fixtures](https://github.com/d-v-b/pydantic-bigstitcher/tree/main/tests/fixture) and adjust the models until everything passes :) Hopefully widening some types into unions gets everything green.

As for additional elements, I would look into the `pydantic-xml` docs to see how extra elements are controlled. In regular pydantic you can specify if extra keys are permitted, but I don't know how this works in `pydantic-xml`.

m-albert commented:
Hey @d-v-b ,

thank you for the quick response here!

> So it's very likely that Nikita's data was processed with a different version of bigstitcher than what I wrote this code to support (the allen institute's data pipeline).

That makes sense.

> my observation has been that bigstitcher xml documents change their schema without warning as new versions are released.

Good to know! Great, if I understand you correctly you're saying it still makes sense to flexibly model this here.

Thanks a lot for the tips regarding the implementation!

> Generally the way to handle this on the pydantic side is to define certain fields as unions over the various possibilities:

Great, done. Had gone the right way in one case but not in the other.

> Operationally, I would recommend adding the full xml document to the set of [test fixtures](https://github.com/d-v-b/pydantic-bigstitcher/tree/main/tests/fixture)

Done.

> Hopefully widening some types into unions gets everything green.

So far parsing works well!

> As for additional elements, I would look into the pydantic-xml docs to see how extra elements are controlled. In regular pydantic you can specify if extra keys are permitted, but I don't know how this works in pydantic-xml.

It's the round trip tests that fail. I've been looking a bit into how to add extra elements with pydantic-xml but haven't found a good way yet. Will get back to this in the next days! Thanks again!

d-v-b commented on src/pydantic_bigstitcher/__init__.py:
```diff
@@ -36,7 +36,7 @@ class ZGroupD(BaseXmlModel, tag="zgroup"):
     path: str = attr()
     tp: str = attr()
-
+    indicies: Optional[str] = attr(default=None)
```
would `str | None` work here?

m-albert replied:
Oh right 👍

m-albert commented:
Hey @d-v-b,

> Operationally, I would recommend adding the full xml document to the set of [test fixtures](https://github.com/d-v-b/pydantic-bigstitcher/tree/main/tests/fixture) and adjust the models until everything passes :) Hopefully widening some types into unions gets everything green.

I added the xml to the test fixtures and everything is green now :)

> As for additional elements, I would look into the pydantic-xml docs to see how extra elements are controlled. In regular pydantic you can specify if extra keys are permitted, but I don't know how this works in pydantic-xml.

It seems that in `pydantic-xml` one can't super easily specify that additional children are permitted, so I used a strategy of handling this manually in `to_xml` and `from_xml`. I added this as a decorator `@with_extra_children` so it can be reused for other classes in case additional children come up again.

>    indices: WhiteSpaceSeparatedStringBecauseArraysDon'tExistLOL

😁 Actually in the new xml it's "indicies" (probably a typo), so I was wondering whether to add more logic to capture additional attributes as well. But then thought that before overcomplicating things for now an optional attribute will also do.

Curious to hear what you think 🙏

m-albert commented:
I had somehow ignored the failing test locally 😅. Instead of an optional "indicies" attribute, the zgroup coming from the mesospim example xml now gets its own zgroup class.

Everything should pass now hopefully.
